### PR TITLE
Some perf improvements for ExpressionStringBuilder

### PR DESCRIPTION
**Pull request by bartdesmet**
This addresses issue #11445.

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 1]
```diff
@@ -217,94 +216,54 @@ protected internal override Expression VisitBinary(BinaryExpression node)
...
+                    case ExpressionType.RightShiftAssign:      op = ">>=";                       break;
+                    case ExpressionType.And:                   op = IsBool(node) ? "And" : "&";  break;
+                    case ExpressionType.AndAssign:             op = IsBool(node) ? "&&=" : "&="; break;
+                    case ExpressionType.Or:                    op = IsBool(node) ? "Or"  : "|";  break;
+                    case ExpressionType.OrAssign:              op = IsBool(node) ? "||=" : "|="; break;
+                    case ExpressionType.ExclusiveOr:           op = "^";                         break;
+                    case ExpressionType.ExclusiveOrAssign:     op = "^=";                        break;
+                    case ExpressionType.Power:                 op = "^";                         break;
+                    case ExpressionType.PowerAssign:           op = "**=";                       break;
+                    case ExpressionType.Coalesce:              op = "??";                        break;
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: Alas, we don't have a `switch` or `match` expression yet, so we need these tedious `break` statements. I've kept the order the same compared to the original, but this should make it easier to match columns in a glimpse. The only change is the conditional in `And`, `Or`, and the compound assignment variants.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 2]
```diff
@@ -354,9 +313,9 @@ protected internal override Expression VisitListInit(ListInitExpression node)
                 {
                     Out(", ");
                 }
-                Out(node.Initializers[i].ToString());
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: This causes allocation of another `ExpressionStringBuilder` instance with a subsequent dispatch into `VisitElementInit` for no particular reason.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 3]
```diff
@@ -632,43 +595,29 @@ protected internal override Expression VisitUnary(UnaryExpression node)
         {
             switch (node.NodeType)
             {
-                case ExpressionType.TypeAs:
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: For the reviewer's convenience, this table is now ordered in the same was as `MakeUnary` over here: https://github.com/dotnet/corefx/blob/8960aac11e908e59573fafbdeff66789db4c338d/src/System.Linq.Expressions/src/System/Linq/Expressions/UnaryExpression.cs#L337

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 4]
```diff
@@ -682,32 +631,24 @@ protected internal override Expression VisitUnary(UnaryExpression node)
...
-                    break;
-                default:
-                    Out(")");
-                    break;
+                case ExpressionType.TypeAs:              Out(" As ");
+                                                         Out(node.Type.Name);
+                                                         Out(')');               break;
+                case ExpressionType.PostIncrementAssign: Out("++");              break;
+                case ExpressionType.PostDecrementAssign: Out("--");              break;
+                default:                                 Out(')');               break;
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: Keeping `default` here; all cases are covered above with `default` throwing for unreachable code. This should cover all cases where we emitted a trailing '(' above, with the exception of `TypeAs` which is treated specially. Please double-check this is indeed the case.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 5]
```diff
@@ -717,21 +658,31 @@ protected internal override Expression VisitDefault(DefaultExpression node)
...
             Out("{ ... } ");
             DumpLabel(node.Target);
-            Out(":");
+            Out(':');
             return node;
         }
 
         protected internal override Expression VisitGoto(GotoExpression node)
         {
-            Out(node.Kind.ToString().ToLower());
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: A lot of hidden allocs here: `Enum.GetName` which allocates a `string[]` for all the names to index into, and a conversion to lower case on top of it.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 6]
```diff
@@ -717,21 +658,31 @@ protected internal override Expression VisitDefault(DefaultExpression node)
...
+            switch (node.Kind)
+            {
+                case GotoExpressionKind.Goto:     op = "goto";     break;
+                case GotoExpressionKind.Break:    op = "break";    break;
+                case GotoExpressionKind.Continue: op = "continue"; break;
+                case GotoExpressionKind.Return:   op = "return";   break;
+                default:
+                    throw new InvalidOperationException();
+            }
+            Out(op);
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: Note that no space is emitted here, making the keyword glued to the target label's name. This is addressed in a separate PR (https://github.com/dotnet/corefx/pull/11427) to judge against the breaking change bar.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs [hunk 7]
```diff
@@ -204,9 +203,9 @@ protected internal override Expression VisitBinary(BinaryExpression node)
             if (node.NodeType == ExpressionType.ArrayIndex)
             {
                 Visit(node.Left);
-                Out("[");
+                Out('[');
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: This change is a very low-hanging fruit. I think it is an evidence that original implementation was never seriously considered as perf-sensitive.
I am just curious - are you running into perf issues with ToString or this is just driven by a code hygiene?

File: src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs (hunk 4 above)
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionStringBuilder.cs]: The switch above (with throwing default) seems complete to me WRT expressions that can be constructed via the factory.

**Comment by VSadov** [comment]
LGTM

**Pull request merged**